docs/contributing/effects-tutorial.md

```diff
@@ -0,0 +1,634 @@
+pure base monad like `Identity`.
+
+The concrete monad is a "transformer" of our base monad (`fused-effects` calls
+these `Carriers`) , which we compose by stacking them on top of each other, and
+we can use a runner to remove it from the stack of monads.  If our base monad is
+`IO`, we can pass that back up to `main`.  If it's a pure base monad, like
+`Identity`, we can use `runIdentity` to simply remove that base monad and get an
+entirely pure value.
+
+When we apply a runner, like `runReader`, that function specifies a _particular_
+carrier monad and removes it from the stack of carriers. Let's take a look at
+some of the definitions needed to make a reader work:
+
+```haskell
+-- This is the Effect type, which all concrete implementors must handle.
+-- Note the GADT syntax, this is almost always required, especially when the
+-- return types are different.
+-- The `k` stands for "kontinuation".  It's a little silly, but it's common.
+data Reader r m k where
+  -- We can think of `Ask` and `Local` as messages sent to some handler later.
+  -- `Ask` takes no extra data, and when handled, returns `Reader r m r`, where
+  -- `r` is the type of the immutable context that the reader holds.
+  -- `Local` needs a function and a monadic action to construct the message.
+  Ask   ::                    Reader r m r
+  Local :: (r -> r) -> m a -> Reader r m a
+
+-- Has is just a convenience wrapper around two typeclasses.
+-- 'Member' doesn't do anything computationally, but it helps prove certain
+-- constraints are met at compile time.  We'll go into more detail later.
+type Has eff sig m = (Member eff sig, Algebra sig m)
+
+-- We define these 'send'-based helpers so that our code isn't completely
+-- cluttered with calls to 'send'.
+ask :: Has (Reader r) sig m => m r
+ask = send Ask
+
+-- 'send' invokes the 'Algebra' typeclass, and only for those carriers
+-- whose 'Member' constraints satisfy the given 'eff' effect.
+-- This is the secret sauce that actually invokes the various 'Algebra'
+-- implementations, where all of the real effect-handling starts.
+send :: Has eff sig m => eff m a -> m a
+
+-- This function specifies a specific datatype, (ReaderC) and "runs" it,
+-- removing ReaderC from the stack of monads.
+runReader :: r -> ReaderC r m a -> m a
+
+-- ReaderC is a simple newtype, which is very common for Carrier types.
+newtype ReaderC r m a = ReaderC (r -> m a)
+
+-- Carriers must be monads, and therefore also applicatives and functors.
+instance Functor      m => Functor      (ReaderC r m)
+instance Applicative  m => Applicative  (ReaderC r m)
+instance Monad        m => Monad        (ReaderC r m)
+  ...
+
+-- Here is where we define the actual operations of the reader, for the specific
+-- carrier ReaderC.
+instance Algebra sig m => Algebra (Reader r :+: sig) (ReaderC r m) where
+  ...
+```
+
+_(Question for critics of this doc: How much of this example code is_
```

Comment:
Hard for me to say.  I know a lot of these things and it still doesn't read easily to me. I think this code uses a lot of extensions that might confuse a newer reader.  It has GADTs, floating instances, type-level operators.

I wonder if we might be better served by a toy example that's simpler.  Those are _much_ harder to write though.